src/audio_buffer.rs

```diff
@@ -0,0 +1,482 @@
+    // if (sample_rate != self.sample_rate()) {
+    //     // @todo - resample
+    //     // cf. https://github.com/HEnquist/rubato/blob/master/examples/fftfixedin64.rs
+    //     // already used in waveshaper
+    // }
+
+    AudioBuffer {
+        number_of_channels,
+        length,
+        // @TODO - this is wrong, should be the AudioContext.sampleRate
+        sample_rate: SampleRate(sample_rate),
+        internal_data: AudioBufferData::from(decoded),
+    }
+}
+
+// `AudioBufferData` is basically a simple matrix representing an aligned,
+// de-interleaved "in memory" audio asset (i.e. a complete sound file), i.e.:
+// 0 (left) : [0.0, 1.0, ...]
+// 1 (right): [0.1, 0.9, ...]
+// ...
+//
+// @note: each channel is an `Arc` that can be cloned by the audio nodes, we
+// only need Arc and not Mutex here because if a node acquired a channel,
+// it must keep a reference to it even if `set_channel_data` is called. In such
+// situation we just replace the channel with a new `Arc` pointing to new `Vec`.
+// So no mutation occurs and memory allocation takes place in the control thread.
+// Having the `Arc` at the level of the single channel, will probably lead to a
+// bit more bookkeeping in the node side, but allows to be a bit more efficent in
+// terms of memory, as we can discard only one channel at a time.
+//
+// @note - spec does not define "audio asset" which can lead to confusion
+//
+// @see - <https://webaudio.github.io/web-audio-api/#acquire-the-content>
+#[derive(Clone)]
+pub(crate) struct AudioBufferData {
+    channels: Rc<Vec<Arc<Vec<f32>>>>,
+}
+
+impl AudioBufferData {
+    // only used by AudioBuffer
+    fn new(number_of_channels: usize, length: usize) -> Self {
+        let mut channels = Vec::<Arc<Vec<f32>>>::with_capacity(number_of_channels);
+        // [spec] Note: This initializes the underlying storage to zero.
+        // we don't want to clone just the Arc here, but the whole underlying data,
+        // so we use `fill_with` instead of `fill`
+        // internal_data.fill_with(|| Arc::new(vec![0.; options.length as usize]));
+        for _ in 0..number_of_channels {
+            let mut channel = Vec::<f32>::with_capacity(length);
+            channel.resize(10, 0.);
+
+            channels.push(Arc::new(channel));
+        }
+
+        Self {
+            channels: Rc::new(channels),
+        }
+    }
+}
+
+// used in `decodeAudioData`
+impl From<Vec<Vec<f32>>> for AudioBufferData {
+    fn from(decoded: Vec<Vec<f32>>) -> Self {
+        let number_of_channels = decoded.len();
+        let mut channels = Vec::<Arc<Vec<f32>>>::with_capacity(number_of_channels);
+
+        for channel in decoded.iter() {
+            // @note - is there a better way than copy using `to_vec`?
+            // basically, is there any way to grab the `decoded.channel` reference and
+            // put it where we want it to be without memory allocation?
+            // (sometimes JS feels so simple... :)
```

Comment:
You can use `decoded.into_iter()` since you have taken ownership. The iterator Item is the actual Vec in that case

Reply:
ok cool